File: backend/app/services/crawler/fetcher.py
"""HTTP fetching with size/time limits, safe redirects and SSRF re-validation."""
from __future__ import annotations

import time
from dataclasses import dataclass, field
from typing import Dict, List, Optional
from urllib.parse import urljoin

import httpx

from app.core.config import settings
from app.services.crawler.url_validator import ValidatedURL, validate_url

MAX_REDIRECTS = 5


class FetchError(Exception):
    def __init__(self, message: str, safe_message: str, kind: str = "fetch_error"):
        super().__init__(message)
        self.safe_message = safe_message
        self.kind = kind


@dataclass
class FetchedPage:
    url: str
    final_url: str
    status_code: int
    headers: Dict[str, str] = field(default_factory=dict)
    content: bytes = b""
    elapsed_ms: int = 0
    history: List[str] = field(default_factory=list)
    error: Optional[str] = None

    @property
    def text(self) -> str:
        if not self.content:
            return ""
        # Prefer the declared charset, fall back to utf-8 with replacement.
        content_type = self.headers.get("content-type", "")
        charset = None
        if "charset=" in content_type.lower():
            charset = content_type.lower().split("charset=", 1)[1].split(";")[0].strip().strip('"')
        if not charset:
            charset = "utf-8"
        try:
            return self.content.decode(charset, errors="replace")
        except LookupError:
            return self.content.decode("utf-8", errors="replace")

    @property
    def size_bytes(self) -> int:
        return len(self.content)


def _build_client() -> httpx.Client:
    return httpx.Client(
        timeout=settings.REQUEST_TIMEOUT_SECONDS,
        follow_redirects=False,
        headers={
            "User-Agent": settings.USER_AGENT,
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
        },
        limits=httpx.Limits(max_connections=settings.CRAWLER_MAX_CONCURRENCY, max_keepalive_connections=2),
    )


def _read_limited(response: httpx.Response) -> bytes:
    """Read a response body up to MAX_RESPONSE_BYTES."""
    limit = settings.MAX_RESPONSE_BYTES
    chunks: List[bytes] = []
    total = 0
    for chunk in response.iter_bytes():
        total += len(chunk)
        if total > limit:
            raise FetchError(
                f"Page exceeds size limit ({total} > {limit} bytes)",
                "This page is too large for the free audit. Try a smaller page.",
                kind="too_large",
            )
        chunks.append(chunk)
    return b"".join(chunks)


def fetch_url(validated: ValidatedURL) -> FetchedPage:
    """Fetch a validated URL. Follows redirects, re-validating every hop."""
    current_url = validated.url
    history: List[str] = []
    started = time.monotonic()

    with _build_client() as client:
        for _ in range(MAX_REDIRECTS + 1):
            try:
                response = client.build_request("GET", current_url)
                resp = client.send(response, stream=True)
            except httpx.TimeoutException as exc:
                raise FetchError(
                    f"Timeout fetching {current_url}: {exc}",
                    "The website took too long to respond.",
                    kind="timeout",
                ) from exc
            except httpx.SSLError as exc:
                raise FetchError(
                    f"SSL error for {current_url}: {exc}",
                    "The website has an SSL certificate problem, so we could not connect securely.",
                    kind="ssl_error",
                ) from exc
            except httpx.ConnectError as exc:
                raise FetchError(
                    f"Connection error for {current_url}: {exc}",
                    "We could not connect to that website. Check the address and try again.",
                    kind="connection_error",
                ) from exc
            except httpx.RequestError as exc:
                raise FetchError(
                    f"Request error for {current_url}: {exc}",
                    "We could not fetch that website.",
                    kind="request_error",
                ) from exc

            if resp.is_redirect:
                location = resp.headers.get("location")
                if not location:
                    raise FetchError("Redirect without Location header", "The website sent an invalid redirect.")
                next_url = urljoin(current_url, location)
                # SSRF guard: re-validate the redirect target before following.
                try:
                    next_validated = validate_url(next_url)
                except Exception as exc:  # URLValidationError
                    raise FetchError(
                        f"Blocked unsafe redirect to {next_url}",
                        "The website redirected us somewhere we cannot safely scan.",
                        kind="unsafe_redirect",
                    ) from exc
                history.append(current_url)
                current_url = next_validated.url
                resp.close()
                continue

            try:
                body = _read_limited(resp)
            finally:
                resp.close()

            elapsed = int((time.monotonic() - started) * 1000)
            return FetchedPage(
                url=validated.url,
                final_url=str(resp.url),
                status_code=resp.status_code,
                headers={k.lower(): v for k, v in resp.headers.items()},
                content=body,
                elapsed_ms=elapsed,
                history=history,
            )

    raise FetchError(
        f"Too many redirects for {validated.url}",
        "The website redirected too many times.",
        kind="too_many_redirects",
    )


def fetch_page(raw_url: str) -> FetchedPage:
    """Validate then fetch. Convenience entrypoint used by the crawler."""
    validated = validate_url(raw_url)
    return fetch_url(validated)

File: backend/src/services/crawler/fetcher.ts
import { config } from "../../config.js";
import { URLValidationError, validateUrl, type ValidatedURL } from "./urlValidator.js";

export const MAX_REDIRECTS = 5;

export class FetchError extends Error {
  safeMessage: string;
  kind: string;
  constructor(message: string, safeMessage: string, kind = "fetch_error") {
    super(message);
    this.name = "FetchError";
    this.safeMessage = safeMessage;
    this.kind = kind;
  }
}

export interface FetchedPage {
  url: string;
  finalUrl: string;
  statusCode: number;
  headers: Record<string, string>;
  content: Buffer;
  elapsedMs: number;
  history: string[];
  error?: string;
}

function headerObject(headers: Headers): Record<string, string> {
  const out: Record<string, string> = {};
  headers.forEach((value, key) => {
    out[key.toLowerCase()] = value;
  });
  return out;
}

function decodeContent(content: Buffer, contentType: string): string {
  let charset = "";
  const lower = contentType.toLowerCase();
  if (lower.includes("charset=")) {
    charset = lower.split("charset=")[1].split(";")[0].trim().replace(/^"|"$/g, "");
  }
  try {
    return new TextDecoder(charset || "utf-8", { fatal: false }).decode(content);
  } catch {
    return new TextDecoder("utf-8", { fatal: false }).decode(content);
  }
}

async function safeCancel(body: ReadableStream<Uint8Array> | null): Promise<void> {
  try {
    await body?.cancel();
  } catch {
    // body may already be consumed/closed; nothing to do
  }
}

async function readLimited(body: ReadableStream<Uint8Array> | null, limit: number): Promise<Buffer> {
  if (!body) return Buffer.alloc(0);
  const reader = body.getReader();
  const chunks: Uint8Array[] = [];
  let total = 0;
  try {
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      total += value.byteLength;
      if (total > limit) {
        await reader.cancel();
        throw new FetchError(
          `Page exceeds size limit (${total} > ${limit} bytes)`,
          "This page is too large for the free audit. Try a smaller page.",
          "too_large"
        );
      }
      chunks.push(value);
    }
  } finally {
    reader.releaseLock();
  }
  return Buffer.concat(chunks);
}

function fetchErrorMessage(err: unknown, url: string): FetchError {
  const anyErr = err as { name?: string; cause?: { code?: string; message?: string }; message?: string };
  const code = anyErr?.cause?.code || anyErr?.name || "";
  const message = anyErr?.cause?.message || anyErr?.message || "";

  if (code === "AbortError" || /timeout|timed out/i.test(message)) {
    return new FetchError(
      `Timeout fetching ${url}: ${message}`,
      "The website took too long to respond.",
      "timeout"
    );
  }
  if (/certificate|CERT_|SSL|TLS|DEPTH_ZERO|UNABLE_TO_VERIFY/i.test(code + message)) {
    return new FetchError(
      `SSL error for ${url}: ${message}`,
      "The website has an SSL certificate problem, so we could not connect securely.",
      "ssl_error"
    );
  }
  if (/ECONNREFUSED|ENOTFOUND|EAI_AGAIN|ECONNRESET|network|fetch failed/i.test(code + message)) {
    return new FetchError(
      `Connection error for ${url}: ${message}`,
      "We could not connect to that website. Check the address and try again.",
      "connection_error"
    );
  }
  return new FetchError(
    `Request error for ${url}: ${message}`,
    "We could not fetch that website.",
    "request_error"
  );
}

export async function fetchUrl(validated: ValidatedURL): Promise<FetchedPage> {
  let currentUrl = validated.url;
  const history: string[] = [];
  const started = Date.now();

  for (let hop = 0; hop <= MAX_REDIRECTS; hop++) {
    let response: Response;
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), config.requestTimeoutSeconds * 1000);
    try {
      response = await fetch(currentUrl, {
        redirect: "manual",
        signal: controller.signal,
        headers: {
          "User-Agent": config.userAgent,
          Accept: "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
          "Accept-Language": "en-US,en;q=0.9",
        },
      });
    } catch (err) {
      throw fetchErrorMessage(err, currentUrl);
    } finally {
      clearTimeout(timer);
    }

    if ([301, 302, 303, 307, 308].includes(response.status)) {
      const location = response.headers.get("location");
      if (!location) {
        await safeCancel(response.body);
        throw new FetchError("Redirect without Location header", "The website sent an invalid redirect.");
      }
      let nextUrl: string;
      try {
        nextUrl = new URL(location, currentUrl).toString();
      } catch {
        await safeCancel(response.body);
        throw new FetchError("Invalid redirect location", "The website sent an invalid redirect.");
      }
      try {
        const nextValidated = await validateUrl(nextUrl);
        history.push(currentUrl);
        currentUrl = nextValidated.url;
        await safeCancel(response.body);
        continue;
      } catch (err) {
        await safeCancel(response.body);
        if (err instanceof URLValidationError) {
          throw new FetchError(
            `Blocked unsafe redirect to ${nextUrl}`,
            "The website redirected us somewhere we cannot safely scan.",
            "unsafe_redirect"
          );
        }
        throw err;
      }
    }

    try {
      const content = await readLimited(response.body, config.maxResponseBytes);
      const elapsedMs = Date.now() - started;
      return {
        url: validated.url,
        finalUrl: response.url || currentUrl,
        statusCode: response.status,
        headers: headerObject(response.headers),
        content,
        elapsedMs,
        history,
      };
    } finally {
      await safeCancel(response.body);
    }
  }

  throw new FetchError(
    `Too many redirects for ${validated.url}`,
    "The website redirected too many times.",
    "too_many_redirects"
  );
}

export async function fetchPage(rawUrl: string): Promise<FetchedPage> {
  const validated = await validateUrl(rawUrl);
  return fetchUrl(validated);
}

export async function probeUrl(url: string): Promise<number | null> {
  try {
    const validated = await validateUrl(url);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), Math.min(config.requestTimeoutSeconds, 10) * 1000);
    try {
      let response = await fetch(validated.url, {
        redirect: "manual",
        signal: controller.signal,
        headers: { "User-Agent": config.userAgent },
        method: "HEAD",
      });
      if (response.status === 405 || response.status === 501) {
        await safeCancel(response.body);
        response = await fetch(validated.url, {
          redirect: "manual",
          signal: controller.signal,
          headers: { "User-Agent": config.userAgent },
          method: "GET",
        });
      }
      const status = response.status;
      await safeCancel(response.body);
      return status;
    } finally {
      clearTimeout(timer);
    }
  } catch {
    return null;
  }
}
